### ISSUE TITLE
Type of class literal incompatible with Callable given corner case dataclass usage

### ISSUE THREAD

**Issue by ktbarrett** [astral-sh/ty#3078]
### Summary

Classes which inherit from a class decorated with `@dataclass`, but who themselves are not decorated with `@dataclass` yield types that are not compatible with Callables that return that class. Adding `@dataclass` works. The types are also compatible with the supertype that is decorated with `@dataclass`.

https://play.ty.dev/9e950b79-447a-46b0-b976-ef302885a7fe

### Version

0.0.23

**Comment by carljm** [comment]
Interesting, definitely looks like a bug, thanks for reporting.

**Comment by carljm** [comment]
Here's a simplified repro:

```py
from dataclasses import dataclass
from typing import Callable

@dataclass(kw_only=True)
class A:
    ...

class B(A):
    ...

@dataclass(kw_only=True)
class C(A):
    ...

b: Callable[[], B] = B  # fails
c: Callable[[], C] = C  # works
```

And here's a repro that eliminates dataclasses and shows the same symptom with a manual class definition:

```py
from typing import Callable

class A:
    def __init__(self: A): ...

class B(A):
    ...

class C(A):
    def __init__(self: C): ...

b: Callable[[], B] = B  # fails
c: Callable[[], C] = C  # works
```

What's happening here is that we currently synthesize a dataclass `__init__` method on `A` using an explicit annotation like `self: A` (though we [have a TODO](https://github.com/astral-sh/ruff/blob/f99b284c1fe1399a82da7f6669467488017d44a7/crates/ty_python_semantic/src/types/class/static_literal.rs#L1348) that maybe we should use `Self` instead).

Then we also use the annotated `self` type to infer the return type, when upcasting the class into a callable type.

(Respecting explicitly-annotated `self` can be important in cases where `__init__` is overloaded, with explicitly annotated `self` type on each overload.)

But for subclasses, this means that `self: A` causes us to consider even `B` as a `Callable[[], A]` rather than a `Callable[[], B]`.

We do seem to handle this better when actually calling the constructor than we do when upcasting to a callable type (these are separate codepaths): `reveal_type(B())` in the above examples reveals `B`, not `A`.

There are several things to explore here:
1. We should probably fix that TODO and use `Self` for synthesized dataclass `__init__` methods -- that would solve the dataclass case, but not the second repro above with explicit `__init__` methods.
2. We should make our upcast-to-callable code consistent with our actually-call-constructor code in terms of how it handles explicit `self` annotations, and avoid this "leakage into subclasses" effect. Probably a subclass should only respect those `self` annotations on `__init__` if they are a specialization of the class being constructed? We should put together some test cases here and see how other type checkers handle them.
